File: java/509_Fibonacci_Number.java
class Solution {
   
    public int fib(int n) {
        if (n == 0) {
            return n;
        }

        int fibNums[] = new int[n + 1];
        fibNums[1] = 1;
        for (int i = 2; i < n + 1; i++) {
            fibNums[i] = fibNums[i - 1] + fibNums[i - 2];
        }
        return fibNums[n];
    }
}

File: php/0500/509_Fibonacci_Number.php
<?php

class Solution {

     /** 
      * @var $memo 
      */
    public $memo = [
        0 => 0,
        1 => 1,
    ];

    /**
     * @param Integer $n
     * @return Integer
     */
    function fib($n) {
        if (!isset($this->memo[$n])) {
            $this->memo[$n] = $this->fib($n - 1) + $this->fib($n - 2);
        }

        return $this->memo[$n];
    }
}

// another iterative approach
//    function fib($n) {
//        if ($n == 0) {
//            return $n;
//        }
//
//        $a = 1;
//        $b = 1;
//        for ($i = 1; $i < $n; $i++) {
//            list($a, $b) = [$b, $a + $b];
//        }
//
//        return $a;
//    }